rosdoc2/verbs/build/builders/sphinx_builder.py

```diff
@@ -53,49 +65,50 @@ def ensure_global(name, default):
-        # Maps markdown files to the "md" lexer, and not the "markdown" lexer
-        # Pygments registers "md" as a valid markdown lexer, and not "markdown"
-        "lexerMapping": {{r".*\.(md|markdown)$": "md",}},
-        # This mapping will work when `exhale` supports `:doxygenpage:` directives
-        # Check https://github.com/svenevs/exhale/issues/111
-        # TODO(aprotyas): Uncomment the mapping below once the above issue is resolved.
-        # "customSpecificationsMapping": utils.makeCustomSpecificationsMapping(
-        #     lambda kind: [":project:", ":path:", ":content-only:"] if kind == "page" else []),
-    }})
+if '{build_type}' != 'ament_python':
+    if rosdoc2_settings.get('enable_breathe', True):
+        # Configure Breathe.
+        # Breathe ingests the XML output from Doxygen and makes it accessible from Sphinx.
+        print('[rosdoc2] enabling breathe', file=sys.stderr)
+        ensure_global('breathe_projects', {{}})
+        breathe_projects.update({{
+    {breathe_projects}}})
+        if breathe_projects:
+            # Enable Breathe and arbitrarily select the first project.
+            extensions.append('breathe')
+            breathe_default_project = next(iter(breathe_projects.keys()))
+
+    if rosdoc2_settings.get('enable_exhale', True):
+        # Configure Exhale.
+        # Exhale uses the output of Doxygen and Breathe to create easier to browse pages
+        # for classes and functions documented with Doxygen.
+        # This is similar to the class hierarchies and namespace listing provided by
+        # Doxygen out of the box.
+        print('[rosdoc2] enabling exhale', file=sys.stderr)
+        extensions.append('exhale')
+        ensure_global('exhale_args', {{}})
+
+        from exhale import utils
+        exhale_args.update({{
+            # These arguments are required.
+            "containmentFolder": "{user_sourcedir}/api",
+            "rootFileName": "library_root.rst",
+            "rootFileTitle": "{package_name} API",
+            "doxygenStripFromPath": "..",
+            # Suggested optional arguments.
+            "createTreeView": True,
+            # TIP: if using the sphinx-bootstrap-theme, you need
+            # "treeViewIsBootstrap": True,
+            "exhaleExecutesDoxygen": False,
+            # Maps markdown files to the "md" lexer, and not the "markdown" lexer
+            # Pygments registers "md" as a valid markdown lexer, and not "markdown"
+            "lexerMapping": {{r".*\.(md|markdown)$": "md",}},
+            # This mapping will work when `exhale` supports `:doxygenpage:` directives
+            # Check https://github.com/svenevs/exhale/issues/111
+            # TODO(aprotyas): Uncomment the mapping below once the above issue is resolved.
+            # "customSpecificationsMapping": utils.makeCustomSpecificationsMapping(
+            #     lambda kind: [":project:", ":path:", ":content-only:"] if kind == "page" else []),
+        }})
```

Comment:
I thought about doing it that way, but once these settings are evaluated (i.e. `breathe`/`exhale` are added to the list of `sphinx` extensions), there is no convenient way to turn off their invocation. We could manually remove these from the list of extensions, conditional on what the build type is, but that would be extra work no?

Reply:
My concern is that a user explicitly turns exhale on (for example) with `rosdoc2_settings['enable_exhale'] = True` but it has no effect because they don't satisfy `'{build_type}' in ('ament_cmake', 'cmake') or {always_run_doxygen}`.

I think the behavior is that we always check the value of `enable_exhale` (for example), but we choose the default (currently always `True`) based on the build type and/or doxygen setting. If the user sets it to `True` manually (overriding the default) and doxygen is not being run we can throw a warning or error for that case.

Reply:
Okay, I see what you mean now. I didn't consider that approach because I thought there was no reason to turn exhale on unless it's an `ament_cmake`/`cmake` build type, but in hindsight you could generate doxygen tags for many types of projects. I will address this in a commit.